middleware/rollup.go

```diff
@@ -0,0 +1,190 @@
+type RollupRouter struct {
+	logger         RollupLogger
+	reportingDelay time.Duration
+	ctx            context.Context
+	ctxDone        bool
+
+	// create a rollup object per unique (status-code, path) pair
+	rollupsMu sync.Mutex
+	rollups   map[string]*logRollup
+}
+
+// NewRollupRouter creates a new log rollup output.
+// Rollups will stop when the context is canceled.
+func NewRollupRouter(ctx context.Context, logger RollupLogger, reportingDelay time.Duration) *RollupRouter {
+	l := &RollupRouter{
+		logger:         logger,
+		reportingDelay: reportingDelay,
+		rollups:        map[string]*logRollup{},
+		ctx:            ctx,
+		ctxDone:        false,
+	}
+	go func() {
+		select {
+		case <-ctx.Done():
+			l.rollupsMu.Lock()
+			l.rollups = map[string]*logRollup{}
+			l.ctxDone = true
+			l.rollupsMu.Unlock()
+		}
+	}()
+	return l
+}
+
+// ShouldRollup returns true when a log msg meets the criteria for rollup.
+func (r *RollupRouter) ShouldRollup(logmsg map[string]interface{}) bool {
+	statusCode, ok := logmsg["status-code"].(int)
+	if !ok {
+		return false
+	}
+
+	responseTime, ok := logmsg["response-time"].(time.Duration)
+	if !ok {
+		return false
+	}
+
+	// in the future allow more configurability, for now default to 200s and < 500ms
+	if statusCode == 200 && responseTime < 500*time.Millisecond {
+		return true
+	}
+
+	return false
+}
+
+// Process rolls up a log message.
+func (r *RollupRouter) Process(logmsg map[string]interface{}) {
+	if r.ctxDone {
+		return
+	}
+
+	statusCode, ok := logmsg["status-code"].(int)
+	if !ok {
+		return
+	}
+	path, ok := logmsg["path"].(string)
+	if !ok {
+		return
+	}
+	canary, ok := logmsg["canary"].(bool)
+	if !ok {
+		return
+	}
+	r.findOrCreate(statusCode, path, canary).add(logmsg)
+}
+
+func (r *RollupRouter) findOrCreate(statusCode int, path string, canary bool) *logRollup {
+	r.rollupsMu.Lock()
+	defer r.rollupsMu.Unlock()
+	rollupKey := fmt.Sprintf("%d-%s", statusCode, path)
+	if canary {
+		rollupKey += "-canary"
+	}
+	if rollup, ok := r.rollups[rollupKey]; ok {
+		return rollup
+	}
+	rollup := &logRollup{
+		Logger:           r.logger,
+		ReportingDelayNs: (r.reportingDelay).Nanoseconds(),
+		StatusCode:       statusCode,
+		Path:             path,
+		Canary:           canary,
+	}
+	r.rollups[rollupKey] = rollup
+	go rollup.schedule(r.ctx)
+	return rollup
+}
+
+// logRollup represents a single rollup.
+type logRollup struct {
+	Logger           RollupLogger
+	ReportingDelayNs int64
+	StatusCode       int
+	Path             string
+	Canary           bool
+
+	rollupMu                sync.Mutex
+	rollupMsg               map[string]interface{}
+	rollupResponseTimeNsSum int64
+}
+
+func (r *logRollup) report() {
+	r.rollupMu.Lock()
+	defer r.rollupMu.Unlock()
+	if r.rollupMsg != nil {
+		r.rollupMsg["response-time-sum"] = r.rollupResponseTimeNsSum
+		r.rollupMsg["response-time"] = r.rollupResponseTimeNsSum / r.rollupMsg["count"].(int64)
```

Comment:
nit: feel free to ignore, but i always wished the response times were in milliseconds instead of nanonseconds so i could read them more easily

Reply:
agreed c9cba52